File: social-media-marketing-machine/frontend/src/types/campaign.ts
/** Mirrors backend/app/schemas/campaign.py::CampaignRead exactly. */
export type CampaignStatus = "draft" | "active" | "paused" | "completed" | "archived";

export interface Campaign {
  id: string;
  organization_id: string;
  client_id: string;
  brand_id: string | null;
  name: string;
  goal: string | null;
  status: CampaignStatus;
  start_date: string | null;
  end_date: string | null;
  budget_cents: number | null;
  created_at: string;
  updated_at: string;
}

/** Mirrors backend/app/schemas/common.py::Page[T]. */
export interface Page<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

File: social-media-marketing-machine/backend/app/schemas/campaign.py
"""Pydantic request/response schemas for the Campaign resource."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import CampaignStatus


class CampaignBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    goal: Optional[str] = Field(default=None, max_length=1000)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    budget_cents: Optional[int] = Field(default=None, ge=0)


class CampaignCreate(CampaignBase):
    client_id: uuid.UUID
    brand_id: Optional[uuid.UUID] = None
    status: CampaignStatus = CampaignStatus.DRAFT


class CampaignUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    goal: Optional[str] = Field(default=None, max_length=1000)
    status: Optional[CampaignStatus] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    budget_cents: Optional[int] = Field(default=None, ge=0)
    brand_id: Optional[uuid.UUID] = None


class CampaignRead(CampaignBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    client_id: uuid.UUID
    brand_id: Optional[uuid.UUID]
    status: CampaignStatus
    created_at: datetime
    updated_at: datetime
